finality-aleph/src/sync/substrate/translator.rs

```diff
@@ -0,0 +1,68 @@
+use std::fmt::{Display, Error as FmtError, Formatter};
+
+use aleph_primitives::BlockNumber;
+use sp_blockchain::Backend;
+use sp_runtime::traits::{Block, Header};
+
+use crate::{
+    justification::AlephJustification,
+    sync::{
+        substrate::{
+            chain_status::{Error as ChainStatusError, SubstrateChainStatus},
+            BlockId, Justification, JustificationTranslator,
+        },
+        BlockStatus, ChainStatus,
+    },
+};
+
+pub enum Error<B: Block> {
+    ChainStatus(ChainStatusError<B>),
+    NoBlock,
+    AlreadyJustified,
+}
+
+impl<B: Block> Display for Error<B> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            ChainStatus(e) => {
+                write!(f, "error retrieving block status: {}", e)
+            }
+            NoBlock => write!(f, "block not present"),
+            AlreadyJustified => write!(f, "block already justified"),
+        }
+    }
+}
+
+impl<B: Block> From<ChainStatusError<B>> for Error<B> {
+    fn from(value: ChainStatusError<B>) -> Self {
+        Error::ChainStatus(value)
+    }
+}
+
+impl<B, BE> JustificationTranslator<B::Header> for SubstrateChainStatus<B, BE>
+where
+    B: Block,
+    B::Header: Header<Number = BlockNumber>,
+    BE: Backend<B>,
+{
+    type Error = Error<B>;
+
+    fn translate(
+        &self,
+        raw_justification: AlephJustification,
+        hash: <B::Header as Header>::Hash,
+        number: BlockNumber,
+    ) -> Result<Justification<B::Header>, Self::Error> {
+        use BlockStatus::*;
+        let block_id = BlockId::new(hash, number);
```

Comment:
nope, currently this is for all Justifications that come from network/rpc etc. - we expect `Error::NoBlock` to be raised in such case

Reply:
Hm, interesting. No, they are not verified in any way, but if you submit a mismatched hash/number pair, then only the hash gets used and the returned header contains the correct number. This is kinda unexpected behaviour, but I think it should be fixed on the level of `SubstrateChainStatus` – it should return an error then, rather than a header/justification.

I think I can do this as part of this PR, unless you object. <_<

Reply:
+1

Reply:
Ok, added appropriate checks to `ChainStatus` and now it returns errors when it gets block IDs that are inconsistent in this way.